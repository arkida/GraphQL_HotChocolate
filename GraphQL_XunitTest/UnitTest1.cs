using Newtonsoft.Json;
using GraphQL_HotChocolate.Model;
using Xunit;
using Xunit.Priority;
using GraphQL_HotChocolate.GraphQL;
using Microsoft.Extensions.DependencyInjection;
using System.Threading.Tasks;
using HotChocolate.Execution;
using Snapshooter.Xunit;
using GreenDonut;
using Moq;
using GraphQL_HotChocolate.Services;
using Microsoft.AspNetCore.Identity;
using System.Reflection;
using System.Net;
using Moq.Protected;

namespace GraphQL_XunitTest
{
    public class UnitTest1
    {

        public UnitTest1()
        {
        }

        [Fact, Priority(1)]
        public async Task GenerateSchema_MatchesExistingSnapshot()
        {
            // Arrange
            var schema = await new ServiceCollection()
                .AddGraphQLServer()
                .AddQueryType<Query>()
                .BuildSchemaAsync();

            // Act & Assert
            schema.ToString().MatchSnapshot();
        }


        [Fact, Priority(2)]
        public void GetSwapiPeople_Test()
        {
            // All requests made with HttpClient go through its handler's SendAsync() which we mock
            var httpMessageHandlerMock = new Mock<HttpMessageHandler>();


            string sCurrentDirectory = AppDomain.CurrentDomain.BaseDirectory;
            string sFile = System.IO.Path.Combine(sCurrentDirectory, @"..\..\..\MockFiles\SwapiPeople_DataResponse.txt");
            string sFilePath = Path.GetFullPath(sFile);
            var jsonData = System.IO.File.ReadAllText(sFilePath);

            var people = JsonConvert.DeserializeObject<SwapiPeople>(jsonData);

            var response = new HttpResponseMessage
            {
                StatusCode = HttpStatusCode.OK,
                Content = new StringContent(jsonData),
            };
            
            // Set up the SendAsync method behavior.
            httpMessageHandlerMock
                .Protected() // <= here is the trick to set up protected!!!
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.IsAny<HttpRequestMessage>(),
                    ItExpr.IsAny<CancellationToken>())
                .ReturnsAsync(response);

            // create the HttpClient
            var httpClient = new HttpClient(httpMessageHandlerMock.Object)
            {
                BaseAddress = new Uri("https://swapi.dev/api/people/")
            };

            var swapiPeopleService = new SwapiPeopleService(httpClient);

            var result = swapiPeopleService.GetSwapiPeople();

            Assert.Equal(people.Origin, result.Result.Origin);

            httpMessageHandlerMock.Protected().Verify(
               "SendAsync",
               Times.Exactly(1),
               ItExpr.Is<HttpRequestMessage>(req => req.Method == HttpMethod.Get),
               ItExpr.IsAny<CancellationToken>());
        }


        [Fact, Priority(3)]
        public void GetSwapiPersonId_Test()
        {
            // All requests made with HttpClient go through its handler's SendAsync() which we mock
            var httpMessageHandlerMock = new Mock<HttpMessageHandler>();

            string sCurrentDirectory = AppDomain.CurrentDomain.BaseDirectory;
            string sFile = System.IO.Path.Combine(sCurrentDirectory, @"..\..\..\MockFiles\SwapiPerson_DataResponse.txt");
            string sFilePath = Path.GetFullPath(sFile);
            var jsonData = System.IO.File.ReadAllText(sFilePath);

            var person = JsonConvert.DeserializeObject<SwapiPerson>(jsonData);

            var response = new HttpResponseMessage
            {
                StatusCode = HttpStatusCode.OK,
                Content = new StringContent(jsonData),
            };

            // Set up the SendAsync method behavior.
            httpMessageHandlerMock
                .Protected() // <= here is the trick to set up protected!!!
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.Is<HttpRequestMessage>(m => m.RequestUri!.Equals("https://swapi.dev/api/people/23")),
                    ItExpr.IsAny<CancellationToken>())
                .ReturnsAsync(response);

            // create the HttpClient
            var httpClient = new HttpClient(httpMessageHandlerMock.Object)
            {
                BaseAddress = new Uri("https://swapi.dev/api/people/")
            };

            var swapiPeopleService = new SwapiPeopleService(httpClient);

            var result = swapiPeopleService.GetSwapiPersonById(person.Id);

            Assert.NotNull(result.Result);

            httpMessageHandlerMock.Protected().Verify(
               "SendAsync",
               Times.Exactly(1),
               ItExpr.Is<HttpRequestMessage>(req => req.Method == HttpMethod.Get),
               ItExpr.IsAny<CancellationToken>());
        }

    }
}