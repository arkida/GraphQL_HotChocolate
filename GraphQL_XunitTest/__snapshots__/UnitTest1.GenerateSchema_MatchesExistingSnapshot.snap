schema {
  query: Query
}

type Query {
  swapiPeople: SwapiPeople!
  swapiPerson(id: Int!): SwapiPerson!
}

type SwapiPeople {
  origin: String!
  count: Int!
  next: String!
  previous: String!
  results: [SwapiPerson!]
  isSuccessful: Boolean!
  code: String!
  message: String!
}

type SwapiPerson {
  id: Int!
  name: String!
  height: String!
  mass: String!
  hairColor: String!
  skinColor: String!
  eyeColor: String!
  birthYear: String!
  gender: String!
  homeworId: String!
  films: [String!]
  species: [String!]
  vehicles: [String!]
  starships: [String!]
  created: String!
  edited: String!
  url: String!
  isSuccessful: Boolean!
  code: String!
  message: String!
}

"The `@defer` directive may be provided for fragment spreads and inline fragments to inform the executor to delay the execution of the current fragment to indicate deprioritization of the current fragment. A query with `@defer` directive will cause the request to potentially return multiple responses, where non-deferred data is delivered in the initial response and data deferred is delivered in a subsequent response. `@include` and `@skip` take precedence over `@defer`."
directive @defer("If this argument label has a value other than null, it will be passed on to the result of this defer directive. This label is intended to give client applications a way to identify to which fragment a deferred result belongs to." label: String "Deferred when true." if: Boolean) on FRAGMENT_SPREAD | INLINE_FRAGMENT

"The `@stream` directive may be provided for a field of `List` type so that the backend can leverage technology such as asynchronous iterators to provide a partial list in the initial response, and additional list items in subsequent responses. `@include` and `@skip` take precedence over `@stream`."
directive @stream("If this argument label has a value other than null, it will be passed on to the result of this stream directive. This label is intended to give client applications a way to identify to which fragment a streamed result belongs to." label: String "The initial elements that shall be send down to the consumer." initialCount: Int! = 0 "Streamed when true." if: Boolean) on FIELD
